### PR TITLE
Revcomp version based on the gcc 8 version

### PR DESCRIPTION
**Pull request by bradcray**
revcomp-blc-gcc8.chpl is a version of reverse-complement from the
CLBG website based on the gcc version 8 that's currently leading the
competition.  For the most part, this is a fairly straightforward
transliteration of that version, though I've replaced the use of
semaphores, condition variables, and the like to coordinate between
tasks with atomics for simplicity/efficiency.

revcomp-blc-gcc8-alg.chpl is a variation that implements a few
different approaches (some of them incorrect) to the major places
where I ended up experimenting with different algorithms, as described
below.  For now, I'm using it to time the memchr()-based implementation
with a spin-loop on the atomic, which seems to be the fastest on our
shootout-like configuration and the most stable on my laptop.  I originally
set up the correctness test to try out all algorithms, but it turns out that
sub_test.py has a bug that prevents this, so for now most are commented
out until I get that bug fixed  (in #21350).

Much to my surprise, a big part of the execution time turned out to be
in the findSeqStart() routine, which I'd originally written serially thinking
that it wouldn't be enough of a bottleneck to be an issue.  This would be
nice to write as an array.find() operation, relying on the array implementation
to do something efficient based on the element type and size (e.g.,
call `memchr()`?  call `memchr()` in parallel?  use a homegrown solution
like I do here?).  That could be used to determine the header size as well,
where here I did that with a serial loop, knowing that it was too small
to bother with my parallel implementtion.

Similarly, I did a homegrown memmove() using a serial block to decide
whether it could be done in parallel or not, and that would
be nice to have some sort of shift routine take care of (though this was
not as much of a bottleneck for this specific computation; more an
example of something that required some cleverness to write correctly
and efficiently, where it'd be nice to have arrays just support it.

The other main compromise here was that it'd be nice to use the
array-based version of readBinary(), but it doesn't currently return
the number of elements read on EOF, which is necessary for a case
like this where we don't know how large the array we're reading needs
to be and are just using it as a buffer.  I ended up using similar
readBinary() / writeBinary() forms everywhere for consistency even
though some of them could also have been written as write()s or
write()s of array slices.

Oh, and it'd be nice to have a more Chapeltastic way of getting an efficient
stdin/stdout for computations like this...

BHarsh raised an interesting question about how this coforall + atomics
idiom might be abstracted into an iterator (in a library say).  That deserves
more thought.

### CODE REVIEW

**Comment by bradcray** [comment]
@benharsh : I think this is ready for review and is a code I'd be comfortable submitting to the site (after which, I might continue to update it based on further experimentation, comments, etc.  I do still need to add it to the performance graphs (and would still like to make a version that compares various ways of expressing the memchr(), but would like to get a reasonable version submitted first.

**Comment by bradcray** [comment]
Food for thought:  I'm seeing highly variable execution times from run to run (by ~25% at times) which makes me wonder whether using a simpler round-robin approach to dealing out the chunks of work would help with that since it's not as though the load balance should vary much.  Of course, there'd still need to be coordination on the output.  But how to express it...?  A `forall` over a strided range isn't quite right because we wouldn't have any guarantee that the right iterations would be scheduled in the right order.  And a `coforall` over the full strided range would create too much parallelism arguably...?  So maybe the current technique is not bad...  but then where is the high variability coming from?  (worse than I think I typically see on my laptop).

**Comment by bradcray** [comment]
Also, annoyingly, my best times between eef7a00cf0f9878efd9f2623638ff9604101542c and 6bf4bb9185a559f0ceb6817fa5a5164573a08e63 (or 6e127724150e8eac38eba998404f8522f6244fa3, which is the HEAD commit on this PR) are off by about 5%.  That commit took out some very clumsy bookkeeping code that I wasn't happy with which is handling the very special case of not finding the "next sequence" at the very front of the buffer.  I've struggled to find a way to express this both cleanly and elegantly, but really thought that this approach would do it since the second condition in the `&&` should only fire once we find a `>`, so three times total in the full run.  But maybe something about it complicates the generated code such that it can't be as efficient?

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    dstFront += 2;
+  }
+}
+
+// see whether there's a sequence start ('>') in 'buff[low..#count]'
+proc findSeqStart(buff, low, count, ref ltOff) {
+  ltOff = max(int);
+  forall i in low..#count with (min reduce ltOff) {
+    if buff[i] == '>'.toByte() && i != 0 {
+      ltOff = i;
```
Comment (ronawho) [inline test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:169]: Think this should be `ltOff = min(i, ltOff)`?

**Comment by bradcray** [comment]
@benharsh : Thinking about the running time variability this morning, I'm realizing that I think my use of compareExchange() is completely overly engineered and that I should just be using an atomic subtraction instead (so maybe I shouldn't have claimed to be all that much better at atomics after AoC).  And the waitFor() may be better as a busy-wait since I know I'm not oversubscribing the tasks.

**Comment by bradcray** [comment]
@ronawho:

> FYI on the shootout-like machine I see ...

Huh...  The version in the release isn't permitted anymore (I think it does a cheat like reading the whole file in one fell swoop?  If you have time, would you time the revcomp3 on this issue:  https://salsa.debian.org/benchmarksgame-team/benchmarksgame/-/issues/526 which should work with main.  Also, are you seeing much variability across runs?

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 1 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:169]: Good catch, I'd been thinking this morning about whether I was just getting away with something here, and think you're right.  Unfortunately, adding the min takes the time down a notch further.

Comment (ronawho) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: Can you iterate in reverse or maybe do some sort of per task eurika?

```chpl
if == max(int) && buff[i] == '>'.toByte() && i != 0
```

or 

```chpl
forall i in low..#count by -1 ...
```

Not sure how optimized parallel negative strided range iterators are though..

**Comment by ronawho** [comment]
> If you have time, would you time the revcomp3 

```
# revcomp3 from https://salsa.debian.org/benchmarksgame-team/benchmarksgame/-/issues/526
time -p ./revcomp3 --n=0 < revcomp-input100000000.txt > /dev/null
# real 1.82
# user 2.43
# sys 0.87
```

> Also, are you seeing much variability across runs?

No, pretty stable

**Comment by ronawho** [comment]
>  And the waitFor() may be better as a busy-wait since I know I'm not oversubscribing the tasks.

I would highly recommend against busy waiting since you'd be relying on implementation details to know that tasks wouldn't be running on the same thread.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 1 above)
Comment (ronawho) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: Hmm, and thinking some more -- it really surprises me that the `min` hurts performance. I was thinking there weren't that many instances of `>` so we wouldn't be calling min that often, but maybe that's wrong.

**Comment by bradcray** [comment]
> time -p ./revcomp3 --n=0 < revcomp-input100000000.txt > /dev/null

OK, that's a little reassuring (in that the new version is still a big improvement even though it's not as competitive with the gcc #8 version as I was hoping).  Still, I wonder why they're almost neck-and-neck on my Mac yet 2x off on this box... :(

> I would highly recommend against busy waiting since you'd be relying on implementation details to know that tasks wouldn't be running on the same thread.

I would think we'd guarantee that a `coforall` with # iterations <= here.maxTaskPar will not schedule the tasks in a way that they can prevent each other from running (though I'll admit we don't make that guarantee today).  Would you disagree?

**Comment by ronawho** [comment]
> I would think we'd guarantee that a coforall with # iterations <= here.maxTaskPar will not schedule the tasks in a way that they can prevent each other from running (though I'll admit we don't make that guarantee today). Would you disagree?

We don't guarantee this and I wouldn't want to constrain our implementation by guaranteeing this.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            lastLineGaps = cols - (lastLineChars + 1);
+
+      var cursor = myStartChar + headerSize,
+          chunkLeft = myChunkSize,
+          chunkPos = 0;
+
+      if lastLineGaps == 0 {
+        revcomp(chunkPos, cursor, chunkLeft, myBuff, seq);
+      } else {
+        while chunkLeft >= cols {
```
Comment (benharsh) [inline test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:112]: I think this else-branch might benefit from a brief comment explaining its purpose. Or a comment somewhere briefly describing the "last line" issue.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+      // then shift the next sequence to the start of the buffer
+      newChars -= nextSeqStart - readPos + 1;
+
+      serial (nextSeqStart < newChars) do
+        forall j in 0..newChars do
+          buff[j] = buff[j+nextSeqStart];
+
+      // and reset to see whether there's another sequence ahead
+      readPos = 1;
```
Comment (benharsh) [inline test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:63]: Should this reset to ``0``? Or maybe something like ``buff.domain.first``?

**Comment by bradcray** [comment]
>  I wouldn't want to constrain our implementation by guaranteeing this.

Do you have cases in mind where you think we'd want to map such a coforall's tasks to the same thread?  I don't see when it would be beneficial.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    // update the position to read to next
+    readPos += newChars;
+
+    // if we're about to run out of space, grow the buffer by 2x
+    if readPos + readSize > buffCap {
+      buffCap *= 2;
+      buffDom = {0..<buffCap};
+    }
+  } while newChars;
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:74]: Yeah, this is an intended-from-the-start feature of Chapel.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 3 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:63]: I've bashed my head against this several times and been unable to come up with a version as clean as this one that passes all my corner case tests, though I feel as though an even cleaner version ought to exist.  The rationale for setting it to 1 here is that we've just shifted `>` to the 0th position, so there's no need to check that position again.  Probably setting it to 0 could also be made to work (since we have to check for `low==0` / `i != 0` within `findSeqStart()` currently anyway), but it would also require doing some +/- 1 logic on `newChars`.  This felt like a six / half-dozen case to me, and this felt like the cleanest working version I found.

I feel as though there ought to be a way to set up the very first call to `findSeqStart()` to remove the need for checking against `low` / `i` within the routine (by essentially setting it to the equivalent of this `1` here).  But so far I haven't been able to find an approach that works and is clean.

I'd be very happy if you/someone were able to crack this, though... I've just been unable to (it was the last thing I tried last night before giving up yet again).

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 1 above)
Comment (benharsh) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: Would it make a difference if ``ltOff`` were a local variable declared inside the function, rather than a ``ref`` argument?

**Comment by bradcray** [comment]
> Still, I wonder why they're almost neck-and-neck on my Mac yet 2x off on this box... :(

Oh shoot, it's because I'm an idiot...  My Mac's `time` doesn't support `-p` and this whole time I've mistakenly only been comparing system times, not user times...  :(    That suggests that there may still be lots of other opportunity for optimization that I've been ignoring...  (:(

(though it's also interesting that the system time is nearly at parity on my Mac, yet still 2x off on the shootout-like box...)

Anyway, I'm going to park this for this morning and look at the user timings in more detail when I can rally again.

**Comment by ronawho** [comment]
> Do you have cases in mind where you think we'd want to map such a coforall's tasks to the same thread? I don't see when it would be beneficial.

For a coforall over the number of cores on an idle system, not really. As a contrived example, maybe an energy optimized runtime that would spawn tasks to the same thread and only moves long running tasks at yield points to new threads to keep CPU utilization low.

On a non idle system, a smart runtime could schedule tasks on the least busy threads.

As a real world example, the old sherwood scheduler would spawn tasks to a single shepherd, and rely on work stealing from other threads to spawn tasks out.

But all of these are pretty contrived or not something I think we really want to do, I just don't like making guarantees about task mappings that may constrain us down the road.

**Comment by bradcray** [comment]
> Oh shoot, it's because I'm an idiot... My Mac's time doesn't support -p and this whole time I've mistakenly only been comparing system times, not user times... :( That suggests that there may still be lots of other opportunity for optimization that I've been ignoring... (:(

Correction... I'm not a complete idiot, I just can't read.  And forgot I was using `zsh` for my shell on this Mac rather than `bash`, s.t. the values printed out are in a different order than I'm used to.  So I'm back to the Chapel version being in the right ballpark as the reference version on my Mac, which might suggest trying to understand better where we're lagging on the shootout-like box, or my trying some other systems, like chapcs...

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 1 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: It doesn't seem to make much of a difference, which matches my intuition since most of the work on ltOff is being done on the task-local reduction variable, which will be a local variable anyway.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 2 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:112]: You may be incorrectly assuming that I actually understand this code... :D  I've put in at least a very high-level description, though.

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 3 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl:63]: I wrestled with this more tonight and got a version that improves on things I think, effectively by moving the check for 0 into the callsite for findSeqStart().

File: test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl (hunk 1 above)
Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: The min() really is impactful on my Mac, though, resulting in 10% overhead...  Makes me want to cry because it was the change that took us from being competitive with the C version to still feeling a ways off from it (though we're obvs. still a ways off from it on the shootout box anyway).

Comment (bradcray) [reply test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl]: Forgot about this comment:

> Can you iterate in reverse or maybe do some sort of per task eurika?
> forall i in low..#count by -1 ...

Since foralls are order-independent, I'm not sure that "iterate in reverse" makes sense for them...

**Comment by bradcray** [comment]
OK, popping back up to the top level:

* @benharsh : I think the new version of revcomp-blc-gcc8.chpl addresses your feedback and makes other minor improvements as well, and is ready for review.  Modulo the shootout timing differences from what I've been seeing (and maybe even despite them), I think it's submission worthy, tonight at least.

* @ronawho: I've added revcomp-blc-gcc8-algs.chpl which supports a bunch of different algorithms that I've been experimenting with to see whether different choices make different impacts on the shootout-like box (e.g., maybe tasking is more expensive, so we shouldn't use so many foralls on these short-ish sequences?).  Note that some of them are not legal (noted with comments), so would be more for the sake of looking at lower bounds or to see the impact of a change or the like...  I think that these choices are largely independent from one another, so you could see what magnitudes of difference you get for changing the searchAlg separately from the shiftAlg rather than worrying about any sort of cross product.  The waitAlg is more for stability than anything.

**Comment by bradcray** [comment]
(BTW, hopefully obvious, but no need to do anything about this over the weekend.  I worked on it last night primarily so I wouldn't think about it all weekend).

**Comment by ronawho** [comment]
Default timings:
```
chpl test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8-algs.chpl --fast
time -p ./revcomp-blc-gcc8-algs --n=0 < revcomp-input100000000.txt > /dev/null
# real 1.10 (+- 0.2)
```

Script to run all algs (could be modified manually or maybe with `timeit` to get avg/min/max, but results seem pretty stable):

<details>

```python
#!/usr/bin/env python3

import subprocess, sys, timeit

variants = {
    'shiftAlg': ['SerForall', 'Forall', 'Foreach', 'For', 'MemMove'],
    'waitAlg':  ['WaitFor', 'Spin'],
    'searchAlg': ['MaxLoc', 'Forall', 'OptForall', 'Foreach', 'For', 'MemChr'],
}

revcomp = 'test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8-algs.chpl'

for k in variants.keys():
    for alg in variants[k]:
        subprocess.run('chpl --fast {} -s{}={}'.format(revcomp, k, alg), shell=True)

        # warmup
        subprocess.run('./revcomp-blc-gcc8-algs --n=0', shell=True,
                       stdin=open('revcomp-input100000000.txt', 'r'), stdout=subprocess.DEVNULL)

        # real run
        start_time = timeit.default_timer()
        subprocess.run('./revcomp-blc-gcc8-algs --n=0', shell=True,
                       stdin=open('revcomp-input100000000.txt', 'r'), stdout=subprocess.DEVNULL)
        end_time = timeit.default_timer()

        kalg = '{}={}'.format(k, alg)
        print("{:20s} time {:.2f}s".format(kalg, end_time - start_time))
    print()
```

</details>

All algs results:
```
chapelu@chapvm-ubu:~/eronagha/chapel$ ./revcomp-algs.py
shiftAlg=SerForall   time 1.09s
shiftAlg=Forall      time 1.09s
shiftAlg=Foreach     time 1.10s
shiftAlg=For         time 1.10s
shiftAlg=MemMove     time 1.10s

waitAlg=WaitFor      time 1.11s
waitAlg=Spin         time 1.09s

searchAlg=MaxLoc     time 1.09s
searchAlg=Forall     time 1.03s
searchAlg=OptForall  time 0.95s
searchAlg=Foreach    time 1.33s
searchAlg=For        time 1.33s
searchAlg=MemChr     time 0.82s
```

**Comment by bradcray** [comment]
Awesome, thanks Elliot!  So it looks like most of the changes that helped on my Mac are pretty close to a wash on the shootout box other than using MemChr (note that `OptForall` isn't a legal choice and that I included it just to show the impact of using `min()` rather than assignment in the forall-based reduction).

Even though I'm tempted to use memchr() due to the significant bump in performance, it feels as though submitting the forall-based version is probably the way to go and then to prioritize adding an array.find() that can use memchr() under the covers for further improvements.  Or we could wait until a pure-Chapel version was accepted and then submit another one using memchr() and see if he notices/balks.  :)

One thing that doesn't quite make sense is that I think the `searchAlg=Forall` case should be identical to the default case currently (since `SerForall` and `WaitFor` are the defaults), so I'm not sure what to make of the 1.03 there...

**Comment by ronawho** [comment]
> searchAlg=Forall case should be identical to the default case currently

I _think_ `MaxLoc` is the default searchAlg.

**Comment by bradcray** [comment]
> I think MaxLoc is the default searchAlg.

It's the default in the `-algs` version, but not in https://github.com/chapel-lang/chapel/blob/a357dd9541aee305dac7135c3512aa9222d18292/test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl#L163-L171

**Comment by mppf** [comment]
> Oh, and it'd be nice to have a more Chapeltastic way of getting an efficient
stdin/stdout for computations like this...

Do you think that simply having a config param to swap between "Make stdin/stdout interleave correctly with C" and "Don't do that" would be enough?

**Comment by bradcray** [comment]
> Do you think that simply having a config param to swap between "Make stdin/stdout interleave correctly with C" and "Don't do that" would be enough?

No, I think we should have a clean(er) way of doing it in the source code.  When semantic choices like this are only apparent if you happen to know what the compilation line is/was, that always strikes me as a poor approach.  In part because I've been bitten by it in the past ("Why doesn't your source code work?  Oh, it only works if you compile it like so").  In part because I think for a large program, it may want to use different notions of the console readers/writers at different times.  And in the specific case of CLBG because we can't specify different compilation arguments for each CLBG entry (and wouldn't want to, even if we could, I think).

**Comment by mppf** [comment]
For sure it is a task for another issue, but it's not clear what that would look like, so I'd be interested in seeing a proposal.

**Comment by bradcray** [comment]
It's not clear to me either, but the current expression of it is a pretty low bar to improve upon:

```chapel
var stdinBin  = openfd(0).reader(iokind.native, locking=false,
                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED)),
    stdoutBin = openfd(1).writer(iokind.native, locking=false,
                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED));
```

**Comment by bradcray** [comment]
Elliot's timings turned out to be reasonably accurate.  On the CLBG site, they turned out to be:
* gcc #8: 0.48 sec (vs. 0.52 above)
* Chapel #8: 0.91 sec  (vs. 1.03 above)

So while using memchr() within an array find may get us a chunk closer, and possibly even in front of the Rust versions (while also reducing code size), there's still another chunk of performance still missing.  I haven't had the chance to log back into our shootout equivalent since Elliot installed emacs to play with it, but will try it on my next shootout evening...
